File: src/main/kotlin/ai/astha/scantheplanet/idea/scanner/ScannerUtils.kt
package ai.astha.scantheplanet.idea.scanner

import java.io.BufferedReader
import java.io.InputStreamReader
import java.nio.charset.Charset
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import kotlin.io.path.isDirectory
import kotlin.io.path.isRegularFile
import kotlin.io.path.name

object ScannerUtils {
    fun collectFiles(repoPath: Path, filters: PathFilters): List<Path> {
        val root = repoPath.normalize()
        if (!root.isDirectory()) return emptyList()
        val matcherInclude = filters.includeGlobs.map { FileSystems.getDefault().getPathMatcher("glob:$it") }
        val matcherExclude = filters.excludeGlobs.map { FileSystems.getDefault().getPathMatcher("glob:$it") }
        val matcherExcludePatterns = filters.excludePatterns.map { FileSystems.getDefault().getPathMatcher("glob:$it") }
        val matcherIncludeOverride = filters.includeOverride.map { FileSystems.getDefault().getPathMatcher("glob:$it") }

        val files = mutableListOf<Path>()
        Files.walk(root).use { stream ->
            stream.filter { it.isRegularFile() }.forEach { file ->
                val relative = root.relativize(file)
                if (!isIncluded(relative, file, filters, matcherInclude, matcherExclude, matcherExcludePatterns, matcherIncludeOverride)) {
                    return@forEach
                }
                files.add(file)
            }
        }
        return files
    }

    fun isIncluded(
        relativePath: Path,
        absolutePath: Path,
        filters: PathFilters,
        includeMatchers: List<java.nio.file.PathMatcher>,
        excludeMatchers: List<java.nio.file.PathMatcher>,
        excludePatternMatchers: List<java.nio.file.PathMatcher>,
        includeOverrideMatchers: List<java.nio.file.PathMatcher>
    ): Boolean {
        val fileName = relativePath.name
        val ext = fileName.substringAfterLast('.', "").lowercase()
        if (isGitPath(relativePath)) return false
        if (isLockfile(fileName)) return false
        if (isIdeMetadataPath(relativePath)) return false
        if (isIgnoredDotfile(fileName)) return false
        if (filters.onlyFiles != null && relativePath !in filters.onlyFiles) return false

        if (filters.excludeDocs && isDocFile(fileName)) return false
        if (filters.excludeTests && isTestFile(relativePath)) return false

        if (filters.excludeExtensions.isNotEmpty() && ext.isNotEmpty() && filters.excludeExtensions.contains(ext)) {
            return false
        }
        if (filters.includeExtensions.isNotEmpty() && ext.isNotEmpty() && !filters.includeExtensions.contains(ext)) {
            return false
        }

        val pathString = relativePath.toString()
        if (excludeMatchers.any { it.matches(relativePath) } || excludePatternMatchers.any { it.matches(relativePath) }) {
            if (includeOverrideMatchers.any { it.matches(relativePath) }) {
                return true
            }
            return false
        }
        if (includeMatchers.isNotEmpty() && includeMatchers.none { it.matches(relativePath) }) {
            return false
        }

        if (filters.maxFileBytes != null) {
            val size = try { Files.size(absolutePath) } catch (_: Exception) { 0L }
            if (size > filters.maxFileBytes) return false
        }

        return true
    }

    private fun isGitPath(path: Path): Boolean {
        val parts = path.toString().split('/', '\\')
        return parts.any { it == ".git" }
    }

    private fun isLockfile(fileName: String): Boolean {
        val lower = fileName.lowercase()
        if (lower.endsWith(".lock")) return true
        return lower in setOf(
            "package-lock.json",
            "pnpm-lock.yaml",
            "yarn.lock",
            "poetry.lock",
            "pipfile.lock",
            "uv.lock",
            "composer.lock",
            "cargo.lock",
            "go.sum",
            "mix.lock",
            "gemfile.lock",
            "podfile.lock"
        )
    }

    private fun isIdeMetadataPath(path: Path): Boolean {
        val parts = path.toString().split('/', '\\')
        return parts.any { it == ".idea" || it == ".vscode" }
    }

    private fun isIgnoredDotfile(fileName: String): Boolean {
        val lower = fileName.lowercase()
        if (!lower.startsWith(".")) return false
        return lower in setOf(
            ".gitignore",
            ".gitattributes",
            ".gitmodules",
            ".python-version",
            ".editorconfig"
        ) || lower.endsWith(".iml")
    }

    fun isBinary(bytes: ByteArray): Boolean = bytes.any { it == 0.toByte() }

    fun readTextIfValid(path: Path, maxLineChars: Int = maxLineChars()): List<String>? {
        val bytes = try { Files.readAllBytes(path) } catch (_: Exception) { return null }
        if (isBinary(bytes)) return null
        val text = try { bytes.toString(Charset.forName("UTF-8")) } catch (_: Exception) { return null }
        val lines = text.lines()
        if (lines.isEmpty()) return null
        if (lines.any { it.length > maxLineChars }) return null
        return lines
    }

    private fun maxLineChars(): Int {
        val raw = System.getenv("MAX_LINE_CHARS") ?: return 10000
        val parsed = raw.toIntOrNull() ?: return 10000
        return parsed.coerceIn(1000, 200000)
    }

    fun isDocFile(fileName: String): Boolean {
        val lower = fileName.lowercase()
        if (lower == "readme" || lower.startsWith("readme.")) return true
        if (lower == "license" || lower.startsWith("license.")) return true
        val ext = lower.substringAfterLast('.', "")
        return ext in setOf("md", "rst", "txt", "adoc")
    }

    fun isTestFile(path: Path): Boolean {
        val lower = path.toString().lowercase()
        return lower.contains("/test/") || lower.contains("/tests/") ||
            lower.contains("__tests__") || lower.contains(".test.") || lower.contains(".spec.")
    }

    fun gitChangedFiles(repoPath: Path, baseRef: String, includeUntracked: Boolean): Set<Path> {
        val files = mutableSetOf<Path>()
        val diff = runGit(repoPath, listOf("diff", "--name-only", baseRef, "HEAD"))
        files.addAll(diff)
        if (includeUntracked) {
            files.addAll(runGit(repoPath, listOf("ls-files", "--others", "--exclude-standard")))
        }
        return files
    }

    fun gitDiffAddedLines(repoPath: Path, baseRef: String): Map<Path, Set<Int>> {
        val diffText = runGitText(repoPath, listOf("diff", "--unified=0", baseRef, "HEAD"))
        if (diffText.isBlank()) return emptyMap()
        return GitDiff.parseUnifiedDiff(repoPath, diffText)
    }

    private fun runGit(repoPath: Path, args: List<String>): Set<Path> {
        return try {
            val process = ProcessBuilder(listOf("git", "-C", repoPath.toString()) + args)
                .redirectErrorStream(true)
                .start()
            val output = BufferedReader(InputStreamReader(process.inputStream)).readText()
            process.waitFor()
            output.lines()
                .map { it.trim() }
                .filter { it.isNotEmpty() }
                .map { repoPath.resolve(it).normalize() }
                .toSet()
        } catch (_: Exception) {
            emptySet()
        }
    }

    private fun runGitText(repoPath: Path, args: List<String>): String {
        return try {
            val process = ProcessBuilder(listOf("git", "-C", repoPath.toString()) + args)
                .redirectErrorStream(true)
                .start()
            val output = BufferedReader(InputStreamReader(process.inputStream)).readText()
            process.waitFor()
            output
        } catch (_: Exception) {
            ""
        }
    }
}

File: packages/core/src/scanner/ScannerUtils.ts
import { readdirSync, readFileSync, statSync } from 'node:fs';
import { join, normalize, relative } from 'node:path';
import { spawnSync } from 'node:child_process';
import type { PathFilters } from './Scope.js';
import { parseUnifiedDiff } from './GitDiff.js';

export interface GlobMatchers {
  include: RegExp[];
  exclude: RegExp[];
  excludePatterns: RegExp[];
  includeOverride: RegExp[];
}

export function collectFiles(
  repoPath: string,
  filters: PathFilters,
  matchers: GlobMatchers = createGlobMatchers(filters)
): string[] {
  const root = normalize(repoPath);
  const files: string[] = [];
  walk(root, (filePath) => {
    const rel = relative(root, filePath);
    if (!isIncluded(rel, filePath, filters, matchers)) {
      return;
    }
    files.push(filePath);
  });
  return files;
}

function walk(root: string, onFile: (path: string) => void): void {
  let entries: ReturnType<typeof readdirSync>;
  try {
    entries = readdirSync(root, { withFileTypes: true });
  } catch {
    return;
  }
  for (const entry of entries) {
    const fullPath = join(root, entry.name);
    if (entry.isDirectory()) {
      walk(fullPath, onFile);
    } else if (entry.isFile()) {
      onFile(fullPath);
    }
  }
}

export function createGlobMatchers(filters: PathFilters): GlobMatchers {
  return {
    include: filters.includeGlobs.map((pattern) => globToRegex(pattern)),
    exclude: filters.excludeGlobs.map((pattern) => globToRegex(pattern)),
    excludePatterns: filters.excludePatterns.map((pattern) => globToRegex(pattern)),
    includeOverride: filters.includeOverride.map((pattern) => globToRegex(pattern)),
  };
}

export function isIncluded(
  relativePath: string,
  absolutePath: string,
  filters: PathFilters,
  matchers: GlobMatchers
): boolean {
  const fileName = basename(relativePath);
  const ext = extension(fileName).toLowerCase();
  if (isGitPath(relativePath)) return false;
  if (isLockfile(fileName)) return false;
  if (isIdeMetadataPath(relativePath)) return false;
  if (isIgnoredDotfile(fileName)) return false;
  if (filters.onlyFiles && !filters.onlyFiles.has(normalizePath(relativePath))) {
    return false;
  }

  if (filters.excludeDocs && isDocFile(fileName)) return false;
  if (filters.excludeTests && isTestFile(relativePath)) return false;

  if (filters.excludeExtensions.length > 0 && ext) {
    if (filters.excludeExtensions.includes(ext)) return false;
  }
  if (filters.includeExtensions.length > 0 && ext) {
    if (!filters.includeExtensions.includes(ext)) return false;
  }

  const normalized = normalizePath(relativePath);
  if (
    matchAny(matchers.exclude, normalized) ||
    matchAny(matchers.excludePatterns, normalized)
  ) {
    if (matchAny(matchers.includeOverride, normalized)) {
      return true;
    }
    return false;
  }
  if (matchers.include.length > 0 && !matchAny(matchers.include, normalized)) {
    return false;
  }

  if (filters.maxFileBytes != null) {
    const size = safeFileSize(absolutePath);
    if (size > filters.maxFileBytes) return false;
  }

  return true;
}

export function readTextIfValid(
  path: string,
  maxLineChars = maxLineCharsEnv()
): string[] | null {
  let buf: Buffer;
  try {
    buf = readFileSync(path);
  } catch {
    return null;
  }
  if (isBinary(buf)) return null;
  let text: string;
  try {
    text = buf.toString('utf8');
  } catch {
    return null;
  }
  const lines = text.split(/\r?\n/);
  if (lines.length === 0) return null;
  if (lines.some((line) => line.length > maxLineChars)) return null;
  return lines;
}

export function isBinary(buffer: Buffer): boolean {
  return buffer.includes(0);
}

function maxLineCharsEnv(): number {
  const raw = process.env.MAX_LINE_CHARS;
  const parsed = raw ? Number.parseInt(raw, 10) : NaN;
  if (!Number.isNaN(parsed)) {
    return Math.min(Math.max(parsed, 1000), 200000);
  }
  return 10000;
}

export function isDocFile(fileName: string): boolean {
  const lower = fileName.toLowerCase();
  if (lower === 'readme' || lower.startsWith('readme.')) return true;
  if (lower === 'license' || lower.startsWith('license.')) return true;
  const ext = extension(lower);
  return ['md', 'rst', 'txt', 'adoc'].includes(ext);
}

export function isTestFile(path: string): boolean {
  const lower = normalizePath(path).toLowerCase();
  return (
    lower.includes('/test/') ||
    lower.includes('/tests/') ||
    lower.includes('__tests__') ||
    lower.includes('.test.') ||
    lower.includes('.spec.')
  );
}

export function gitChangedFiles(
  repoPath: string,
  baseRef: string,
  includeUntracked: boolean
): Set<string> {
  const files = new Set<string>();
  for (const file of runGit(repoPath, ['diff', '--name-only', baseRef, 'HEAD'])) {
    files.add(join(repoPath, file));
  }
  if (includeUntracked) {
    for (const file of runGit(repoPath, [
      'ls-files',
      '--others',
      '--exclude-standard',
    ])) {
      files.add(join(repoPath, file));
    }
  }
  return files;
}

export function gitDiffAddedLines(
  repoPath: string,
  baseRef: string
): Map<string, Set<number>> {
  const diffText = runGitText(repoPath, [
    'diff',
    '--unified=0',
    baseRef,
    'HEAD',
  ]);
  if (!diffText.trim()) return new Map();
  return parseUnifiedDiff(repoPath, diffText);
}

function runGit(repoPath: string, args: string[]): string[] {
  try {
    const result = spawnSync('git', ['-C', repoPath, ...args], {
      encoding: 'utf8',
    });
    if (result.status !== 0) return [];
    return result.stdout
      .split(/\r?\n/)
      .map((line) => line.trim())
      .filter(Boolean);
  } catch {
    return [];
  }
}

function runGitText(repoPath: string, args: string[]): string {
  try {
    const result = spawnSync('git', ['-C', repoPath, ...args], {
      encoding: 'utf8',
    });
    if (result.status !== 0) return '';
    return result.stdout || '';
  } catch {
    return '';
  }
}

function safeFileSize(path: string): number {
  try {
    return statSync(path).size;
  } catch {
    return 0;
  }
}

function basename(path: string): string {
  const normalized = normalizePath(path);
  const idx = normalized.lastIndexOf('/');
  return idx >= 0 ? normalized.slice(idx + 1) : normalized;
}

function extension(fileName: string): string {
  const idx = fileName.lastIndexOf('.');
  return idx >= 0 ? fileName.slice(idx + 1) : '';
}

function isGitPath(path: string): boolean {
  return normalizePath(path).split('/').some((part) => part === '.git');
}

function isIdeMetadataPath(path: string): boolean {
  return normalizePath(path).split('/').some((part) => part === '.idea' || part === '.vscode');
}

function isIgnoredDotfile(fileName: string): boolean {
  const lower = fileName.toLowerCase();
  if (!lower.startsWith('.')) return false;
  return (
    [
      '.gitignore',
      '.gitattributes',
      '.gitmodules',
      '.python-version',
      '.editorconfig',
    ].includes(lower) || lower.endsWith('.iml')
  );
}

function isLockfile(fileName: string): boolean {
  const lower = fileName.toLowerCase();
  if (lower.endsWith('.lock')) return true;
  return [
    'package-lock.json',
    'pnpm-lock.yaml',
    'yarn.lock',
    'poetry.lock',
    'pipfile.lock',
    'uv.lock',
    'composer.lock',
    'cargo.lock',
    'go.sum',
    'mix.lock',
    'gemfile.lock',
    'podfile.lock',
  ].includes(lower);
}

function normalizePath(path: string): string {
  return path.replace(/\\/g, '/');
}

function matchAny(patterns: RegExp[], value: string): boolean {
  return patterns.some((pattern) => pattern.test(value));
}

function globToRegex(glob: string): RegExp {
  let pattern = normalizePath(glob);
  let out = '^';
  for (let i = 0; i < pattern.length; i += 1) {
    const ch = pattern[i];
    if (ch === '*') {
      const next = pattern[i + 1];
      if (next === '*') {
        out += '.*';
        i += 1;
      } else {
        out += '[^/]*';
      }
      continue;
    }
    if (ch === '?') {
      out += '.';
      continue;
    }
    if ('\\^$+?.()|{}[]'.includes(ch)) {
      out += `\\${ch}`;
      continue;
    }
    out += ch;
  }
  out += '$';
  return new RegExp(out);
}
